src/listeners/ticket/create-modal-submit-validator.ts

```diff
@@ -0,0 +1,183 @@
+import { ApplyOptions } from '@sapphire/decorators';
+import { Events, Listener } from '@sapphire/framework';
+import {
+    ActionRowBuilder,
+    ButtonBuilder,
+    ButtonStyle,
+    ChannelType,
+    EmbedBuilder,
+    GuildMember,
+    PermissionFlagsBits,
+    type Interaction,
+    type ModalSubmitInteraction,
+} from 'discord.js';
+import { Color } from '../../lib/embeds';
+import { TButtonId, InputId, ModalId } from '../../lib/ticket';
+import { messageLogger } from '../..';
+import { settingsManager } from '../..';
+import { userUrl } from '../../lib/user';
+import { ticketLogs } from '../..';
+import type { TicketLog } from '../../lib/ticket';
+
+@ApplyOptions<Listener.Options>({ event: Events.InteractionCreate })
+export class InteractionCreateListener extends Listener<
+    typeof Events.InteractionCreate
+> {
+
+    public override async run(interaction: Interaction) {
+        if (
+            !interaction.isModalSubmit() ||
+            interaction.customId !== ModalId.Ticket ||
+            !interaction.inGuild()
+        ) {
+            return;
+        }
+
+        await interaction.deferReply({ ephemeral: true });
+
+        const title = interaction.fields.getTextInputValue(InputId.Title).trim();
+        if (!title) {
+            return this.sendValidationFailure(
+                interaction,
+                'Title must contain at least 1 non-whitespace character'
+            );
+        }
+
+        const exp = interaction.fields.getTextInputValue(InputId.Explanation).trim();
+        if (!exp) {
+            return this.sendValidationFailure(
+                interaction,
+                'Description must contain at least 1 non-whitespace character'
+            );
+        }
+
+        const guildSettings = await settingsManager.get(interaction.guildId);
+        const guild = await interaction.client.guilds.fetch(interaction.guildId);
+        const member = await guild.members.fetch(interaction.member.user.id);
+
+        const ticketChannelId = guildSettings?.ticketChannel;
+        if (!ticketChannelId) {
+            return;
+        }
+        const ticketChannel = await guild.channels.fetch(
+            ticketChannelId
+        );
+        if (ticketChannel?.type !== ChannelType.GuildText) {
+            return;
+        }
+
+        const latestTicket = await ticketLogs.findOne(
```

Comment:
This method of obtaining the latest ticket number from the database, then incrementing that number, before later saving a new ticket in the database with the incremented number, is open to a race condition. If two tickets were to be created at close to the same time, it would be possible for two tickets to end up with the same number.

Instead, we could change `GuildSettings` in `src/lib/settings.ts` to add a field:
```ts
  lastTicketNumber?: number;
```
And export the `guildSettings` collection defined in `src/index.ts`:
```ts
export const guildSettings = database.collection<GuildSettings>('settings');
```
Then, in this file, we can import `guildSettings as guildSettingsCollection` (to avoid the naming conflict with the local variable `guildSettings` above), and change this code to something similar to the following:
```ts
    const newGuildSettings = await guildSettingsCollection.findOneAndUpdate(
      {_id: interaction.guildId},
      {$inc: {lastTicketNumber: 1}},
      {returnDocument: 'after', upsert: true}
    );
    if (!newGuildSettings.value) {
      return;
    }

    const ticketNumber = newGuildSettings.value.lastTicketNumber!;
    const ticketID = ticketNumber.toString().padStart(6, '0');
```
This code leverages the database's atomic upsert capability to prevent a race condition on the ticket numbers from being possible.